File: app/java-kotlin/src/main/kotlin/com/example/vault/client/VaultConfig.kt
package com.example.vault.client

import com.typesafe.config.ConfigFactory
import io.github.microutils.kotlin.logging.KotlinLogging
import java.time.Duration

private val log = KotlinLogging.logger {}

// TypeSafe Config를 사용하여 설정 로드
data class VaultConfig(
    val vaultAddr: String,
    val namespace: String,
    val roleId: String,
    val secretId: String,
    val kvMountPath: String,
    val kvSecretsPaths: List<String>,
    val renewalInterval: Duration,
    val renewalThresholdRatio: Double
) {
    companion object {
        fun load(): VaultConfig {
            log.info { "⏳ 설정 파일 (application.properties) 로드 시작..." }
            val config = ConfigFactory.load().getConfig("vault")

            val vaultConfig = VaultConfig(
                vaultAddr = config.getString("addr"),
                namespace = config.getString("namespace"),
                roleId = config.getString("role_id"),
                secretId = config.getString("secret_id"),
                kvMountPath = config.getString("kv_mount_path"),
                kvSecretsPaths = config.getStringList("kv_secrets_paths"),
                renewalInterval = Duration.ofSeconds(config.getLong("renewal_interval_seconds")),
                renewalThresholdRatio = config.getInt("token_renewal_threshold_percent") / 100.0
            )

            log.info { "✅ 설정 파일 로드 완료. Vault Address: ${vaultConfig.vaultAddr}" }
            return vaultConfig
        }
    }
}

File: app/net(docker)/VaultConfig.cs
using System.Collections.Generic;
using System;

namespace NewVaultClientDotnet;

public class VaultConfig
{
    // Vault 접속 정보 (string.Empty로 초기화하여 설정 파일 로드를 강제)
    public string Addr { get; set; } = string.Empty;
    public string Namespace { get; set; } = string.Empty; 
    public string RoleId { get; set; } = string.Empty;
    public string SecretId { get; set; } = string.Empty;

    // KV Secret 설정
    public string KvMountPath { get; set; } = string.Empty;
    // 빈 배열로 초기화. KvSecretsPaths 항목이 없으면 빈 배열이 바인딩됩니다.
    public string[] KvSecretsPaths { get; set; } = Array.Empty<string>(); 
    
    // 스케줄링 설정 (0으로 초기화)
    public int RenewalIntervalSeconds { get; set; } = 0; 
    public double TokenRenewalThresholdPercent { get; set; } = 0.0; 
}